GiniVision/Classes/Core/FilePickerManager.swift

```diff
@@ -57,70 +66,51 @@ internal final class FilePickerManager: NSObject {
     
     // MARK: File data picked from gallery or document pickers
     
-    fileprivate func filePicked(withData data: Data) {
+    fileprivate func createDocument(fromData data: Data) -> GiniVisionDocument? {
         let documentBuilder = GiniVisionDocumentBuilder(data: data, documentSource: .external)
         documentBuilder.importMethod = .picker
         
-        if let document = documentBuilder.build() {
-            didPickFile(document)
-        }
+        return documentBuilder.build()
     }
     
-    // MARK: Photo library permission
-    
-    fileprivate func checkPhotoLibraryAccessPermission(deniedHandler: @escaping (_ error: GiniVisionError) -> Void,
-                                                       authorizedHandler: @escaping (() -> Void)) {
-        
-        switch PHPhotoLibrary.authorizationStatus() {
-        case .authorized:
-            authorizedHandler()
-        case .denied, .restricted:
-            deniedHandler(FilePickerError.photoLibraryAccessDenied)
-        case .notDetermined:
-            PHPhotoLibrary.requestAuthorization { status in
-                DispatchQueue.main.async {
-                    if status == PHAuthorizationStatus.authorized {
-                        authorizedHandler()
-                    } else {
-                        deniedHandler(FilePickerError.photoLibraryAccessDenied)
-                    }
-                }
-            }
+    fileprivate func data(fromUrl url: URL) -> Data? {
+        do {
+            _ = url.startAccessingSecurityScopedResource()
+            let data = try Data(contentsOf: url)
+            url.stopAccessingSecurityScopedResource()
+            return data
+        } catch {
+            url.stopAccessingSecurityScopedResource()
         }
+        
+        return nil
     }
     
 }
 
-// MARK: UIImagePickerControllerDelegate, UINavigationControllerDelegate
+// MARK: GalleryCoordinatorDelegate
 
-extension FilePickerManager: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
-    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String: Any]) {
-        if let pickedImage = info[UIImagePickerControllerOriginalImage] as? UIImage,
-            let imageData = UIImageJPEGRepresentation(pickedImage, 1.0) {
-            filePicked(withData: imageData)
-        }
+extension FilePickerManager: GalleryCoordinatorDelegate {
+    func gallery(_ coordinator: GalleryCoordinator, didSelectImageDocuments imageDocuments: [GiniImageDocument]) {
+        didPickDocuments(imageDocuments)
         
-        picker.dismiss(animated: true, completion: nil)
+        coordinator.rootViewController.dismiss(animated: true, completion: nil)
```

Comment:
I'm not sure about this. You expose some coordinator details. That it has a root controller and that it presented something. I'd rather have a method that tells the coordinator that the picker is no longer needed.

Reply:
The `rootViewController` is always exposed since is part of the `Coordinator` protocol. So it cannot be hidden.
Do you have a better idea?

Reply:
That's not the point. It's fine to have the root controller if it support another use case.
The problem I have is that you also expose (and "hardcode") that the coordinator presents view controllers (as opposed to pushing them or embedding them). You also mess with the class internals. It's the age-old discussion between asking a class and telling a class what to do.

So, I'd prefer to have "Tell the coordinator to remove the view controller", rather than "Ask the coordinator for its root controller and tell that root controller to dismiss its presented controller".

Reply:
Got it now. Yes, it makes sense what you say.